feat(curriculum): add http and web standards review page (#67334)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/review-http-and-the-web-standards-model/6a02da945706a3ddc2e44348.md

title: HTTP and the Web Standards Model Review
challengeType: 31
dashedName: review-http-and-the-web-standards-model
---

# --description--

## Servers and the Client-Server Model

- **Server**: A computer or program that receives requests from other computers and sends data or services back over the network. The term refers to both the physical hardware and the software running on it.
- **Client-Server Model**: A distributed architecture where clients send requests and servers handle those requests and return responses.
- **Client**: The process or application (for example, a web browser) that initiates requests to a server.
- **Request-Response Cycle**: The client sends a request; the server processes it and returns a response.

### Server Types

- **Web servers**: Deliver web content (HTML, images, CSS) to browsers.
- **Application servers**: Handle business logic and process user input.
- **Database servers**: Store, retrieve, and manage data.
- **Mail servers**: Send, receive, and store email.
- **File servers**: Manage file storage and sharing across a network.
- **Proxy servers**: Act as intermediaries between clients and other servers.

## DNS (Domain Name System)

> - **DNS**: A hierarchical and decentralized naming system that translates human-readable domain names into IP addresses. Often called the "phone book" of the internet.

Reviewer: ```suggestion
- **DNS**: A hierarchical and decentralized naming system that translates human-readable domain names into IP addresses, often called the "phone book" of the internet.
```

### DNS Resolution Process

1. The browser sends the domain name to the **Recursive Resolver** (usually managed by the ISP).
2. The Recursive Resolver queries **Root Servers**, which direct it to the relevant TLD server.
3. The **TLD Server** returns the address of the **Authoritative Name Server** for the domain.
4. The Authoritative Name Server returns the **DNS Record** containing the IP address.
5. The Recursive Resolver delivers the IP address to the client and **caches** the record to speed up future lookups.

## TCP/IP

- **Protocol**: A set of rules and standards that allow devices to communicate over a network.
- **TCP/IP**: A suite of protocols that determines how data is packaged, sent, routed, and received.
  - **IP (Internet Protocol)**: Routes data packets from source to destination using IP addresses. It is connectionless — each packet travels independently with no delivery guarantee.
  - **TCP (Transmission Control Protocol)**: Breaks data into numbered segments, reassembles them at the destination, checks for errors and lost packets, and requests retransmission if needed. Ensures reliable, ordered delivery.

> - **HTTPS**: A secure version of HTTP where data is encrypted using SSL/TLS before transmission.

Reviewer: This should be part of HTTP section(added a comment there).

```suggestion
```

## HTTP (HyperText Transfer Protocol)

- **HTTP**: An application-level protocol that defines the rules for communication between clients and servers on the web.

> 

Reviewer: ```suggestion

- **HTTPS**: A secure version of HTTP where data is encrypted using SSL/TLS before transmission.

```

- **Headers**: Metadata about the response (for example, `Content-Type: text/html`).
- **Body** (optional): The returned content or error explanation.

```md
HTTP/1.1 200 OK
Content-Type: text/html
Content-Length: 105
Connection: keep-alive

<!DOCTYPE html>
<html>
<head><title>Home</title></head>
<body><h1>Welcome!</h1></body>
</html>
```

## HTTP Status Codes

Status codes are grouped by their first digit:

- **1xx** — Informational responses.
- **2xx** — Successful responses.
- **3xx** — Redirection messages.
- **4xx** — Client error responses.
- **5xx** — Server error responses.

### Common Codes

| Code | Name | Meaning |
| --- | --- | --- |
| `200` | OK | Request succeeded; content returned. |
| `201` | Created | Resource created successfully. |
| `204` | No Content | Success but no content to return. |
| `301` | Moved Permanently | Resource permanently moved to a new URL. |
| `302` | Found | Resource temporarily moved to another URL. |
| `304` | Not Modified | Resource unchanged since last request; use cached version. |
| `400` | Bad Request | Server couldn't understand the request (invalid syntax or data). |
| `401` | Unauthorized | Authentication credentials required. |

> | `403` | Forbidden | Authenticated but lacks permission for the resource. |

Reviewer: 403 does not always mean the user is authenticated, so this can be rephrased:
```suggestion
| `403` | Forbidden | Client lacks permission for the resource, even with valid credentials. |
```

- Each input needs a `name` attribute so the server can identify the data.
- **Client-side validation**: Use the `required` and `pattern` attributes for immediate feedback.
- **Server-side validation**: Always validate on the server for security, regardless of client-side checks.
- The `action` attribute specifies where to send the form data.
- The `method` attribute specifies how to send it (`GET` or `POST`; `PUT`, `PATCH`, `DELETE` require workarounds such as the Fetch API or `method-override` in Express).

### Form Submission Flow

1. User fills out and submits the form.
2. Browser runs client-side validation; proceeds if valid.
3. Browser sends form data to the URL in `action` using the specified `method`.
4. Server receives and processes the data (store in database, authenticate user, etc.).
5. Server may run additional server-side validation.
6. Server responds with a success message, an error, or a redirect.
7. User sees feedback based on the server's response.

## Web Standard Model and Standards Bodies

- **Web standards**: The rules and specifications that browsers follow to render pages and enable web capabilities.

> - **WHATWG**: The standard body that maintains HTML and the DOM as a continuously updated "living standard" driven by major browser engine teams.

Reviewer: ```suggestion
- **WHATWG**: The standards body that maintains HTML and the DOM as a continuously updated "living standard" driven by major browser engine teams.```
